File: portrefs/ion/entities/bullet/GeometricBulletType.java
package ion.entities.bullet;

import arc.util.*;
import arc.math.*;
import arc.graphics.*;
import arc.graphics.g2d.*;
import mindustry.gen.*;
import mindustry.type.*;
import mindustry.game.*;
import mindustry.entities.*;
import mindustry.entities.bullet.*;
import mindustry.graphics.*;

public class GeometricBulletType extends BulletType{
    /** Amount of minimum/maximum sides the bullet can have. */
    public int minSides = 3, maxSides = 7;
    /** Spin speed of bullet. Set to 0 if you don't want it to spin. */
    public float spinSpeed = 2f;
    /** Bullet color. */
    public Color color = Color.white;
    /** Whether or not the bullet is hollow. */
    public boolean hollow = false;
    
    public GeometricBulletType(int min, int max){
        super();
        minSides = min;
        maxSides = max;
    }
    
    public GeometricBulletType(){
        super();
    }
    
    @Override
    public void draw(Bullet b){
        super.draw(b);
        Draw.z(Layer.bullet);
        Draw.color(color);
        
        if(!(b.data() instanceof Integer)) return;
        
        if(hollow){
            Lines.poly(b.x, b.y, ((int)b.data()), b.type.hitSize, b.rotation() + (Time.time * spinSpeed));
        } else {
            Fill.poly(b.x, b.y, ((int)b.data()), b.type.hitSize, b.rotation() + (Time.time * spinSpeed));
        }
    }
    
    @Override
    public Bullet create(Entityc owner, Team team, float x, float y, float angle, float damage, float velocityScl, float lifetimeScl, Object data, Mover mover, float aimX, float aimY){
        Bullet bullet = super.create(owner, team, x, y, angle, damage, velocityScl, lifetimeScl, data, mover, aimX, aimY);
        bullet.data(Mathf.random(minSides, maxSides));
        
        return bullet; //java
    }
}

File: src/ion/entities/bullet/GeometricBulletType.kt
package ion.entities.bullet

import arc.graphics.Color
import arc.graphics.g2d.Draw
import arc.graphics.g2d.Fill
import arc.graphics.g2d.Lines
import arc.math.Mathf
import arc.util.Time
import mindustry.entities.Mover
import mindustry.entities.bullet.BulletType
import mindustry.game.Team
import mindustry.gen.Bullet
import mindustry.gen.Entityc
import mindustry.graphics.Layer

open class GeometricBulletType(min: Int, max: Int) : BulletType() {
    
    var minSides = min
    var maxSides = max
    var spinSpeed = 0.08f
    var color = Color.white
    var hollow = false

    override fun draw(b: Bullet){
        super.draw(b)
        Draw.z(Layer.bullet)
        Draw.color(color)
        //Bullet#data is public but i dont feel like taking risks
        val data = b.data()
        
        if(data is Int){
            if(hollow){
                Lines.poly(b.x, b.y, data, b.type.hitSize, b.rotation() + (Time.time * spinSpeed))
                return
            }
            Fill.poly(b.x, b.y, data, b.type.hitSize, b.rotation() + (Time.time * spinSpeed))
        }
    }
    
    override fun create(
    owner: Entityc?,
    team: Team?,
    x: Float, y: Float,
    angle: Float,
    damage: Float,
    velocityScl: Float,
    lifetimeScl: Float,
    data: Any?,
    mover: Mover?,
    aimX: Float, aimY: Float
    ) = super.create(owner, team, x, y, angle, damage, velocityScl, lifetimeScl, data, mover, aimX, aimY).also{
        it.data(Mathf.random(minSides, maxSides))
    }
}